src/util/kaldi-table-inl.h

```diff
@@ -1631,25 +1751,49 @@ class RandomAccessTableReaderScriptImpl:
         return true;  // we have the key.
       } else {  // preload specified, so we have to pre-load the object before
               // returning true.
-        if (!input_.Open(script_[key_pos].second)) {
-          KALDI_WARN << "Error opening stream "
-                     << PrintableRxfilename(script_[key_pos].second);
-          return false;
+        std::string data_rxfilename; // We will split filename 
+        // script_[key_pos].second into data_rxfilename (e.g. 1.ark:100) 
+        // and range_(if any, e.g. [1:2,2:10]).
+        
+        if (script_[key_pos].second[script_[key_pos].second.size()-1] == ']') {
+          if(!ExtractRangeSpecifier(script_[key_pos].second,
+                                    &data_rxfilename,
+                                    &range_)) {
+            KALDI_WARN << "TableReader: fail to parse file name "
+                       << PrintableRxfilename(script_[key_pos].second);
+            return false;
+          }
         } else {
-          // Make sure holder empty.
-          if (state_ == kHaveObject || state_ == kGaveObject)
-            holder_.Clear();
-          if (holder_.Read(input_.Stream())) {
-            state_ = kHaveObject;
-            current_key_ = key;
-            return true;
-          } else {
-            KALDI_WARN << "Error reading object from "
-                "stream " << PrintableRxfilename(script_[key_pos].second);
-            state_ = kNotHaveObject;
+          data_rxfilename = script_[key_pos].second;
+          if (!range_.empty()) {
+            range_ = "";
+            range_holder_.Clear(); 
+          }
+        }
+        // we only need to update holder_ if the new data_rxfilename is
+        // different from the current data_rxfilename we have.
+        if (data_rxfilename != data_rxfilename_) {
+          if (!input_.Open(data_rxfilename)) {
+            KALDI_WARN << "Error opening stream "
+                       << PrintableRxfilename(data_rxfilename);
             return false;
+          } else {
+            data_rxfilename_ = data_rxfilename;
+            if (holder_.Read(input_.Stream())) {
+              state_ = kHaveObject;
+              current_key_ = key;
+              return true;
+            } else {
+              KALDI_WARN << "Error reading object from "
+                  "stream " << PrintableRxfilename(script_[key_pos].second);
+              state_ = kNotHaveObject;
```

Comment:
You need to check every code path and check that the invariances hold-- e.g. here, we have a code path where holder_contains no object, but data_rxfilename_ is nonempty.